third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/index.move

```diff
@@ -0,0 +1,270 @@
+//# publish
+module 0x42::test {
+
+    struct R has key, drop { value: bool }
+
+    const CONSTANT_VEC: vector<u8> = vector[1, 2, 3];
+
+    fun init(s: &signer) {
+        move_to(s, R{value: true});
+    }
+
+    fun test_resource_1() acquires R {
+        use 0x42::test;
+        assert!(test::R[@0x1].value == true, 0);
+        R[@0x1] = R{value: false};
+        assert!(test::R[@0x1].value == false, 0);
+    }
```

Comment:
@rahxephon89 Are those two tests roughtly equal in the operations performed? Is the `R[@0x1] = R{value: false};` is sugar over `(&mut R[@0x1]).value = false`?

Reply:
yes

Reply:
You can use `R[@0x1].value = false` as well.

Reply:
Isn't that too much sugar? I honestly thought the `R[@0x1] = R { value: false }` equals to the `move_to(@0x1, R { value: false } )` first, before I remembered that the first parameter is a `&signer`. 

If the goal is to remove boilerplate from the `borrow_global` and `borrow_global_mut` calls, I'd suggest instead to go with 
```
borrow_global<Res>(@0x1) -> Res[@0x1]
borrow_global_mut<Res>(@0x1) -> Res[mut @0x1]
```

We're special casing things anyway, and `mut` without ref never used in Move, so it could be easy distinguished. And it won't mess the prioritisation rules, so you could write  

```
Res[mut @0x1].value = false
```

without any parens.

Reply:
> You can use R[@0x1].value = false as well.

Oh, so the compiler knows that we're trying to set a value, and automatically replaces `borrow_global` with `borrow_global_mut` in the translator?

Reply:
For instance, Move already supports
```
let x = R {value: false};
x.value = true; // x is automatically translated to &mut x so that we can change the value.
```
So I think this design aligns with the current design of Move

Reply:
Similarly, for a vector `v`, we can use `v[0] = 1` to update the value.

Reply:
Let me describe it again, so I'd know I understand it correctly (I'm implementing the IDE version of the index expr now):

`R[@0x1]` initially is just a generic "view" on the resource in the global storage. If it's used to read/write it's fields, it becomes either `&` or `&mut` depends on the operation. 
So one can do:
```
// &R 
1 + R[@0x1].value;
// &mut R 
R[@0x1].value = 1;
```
This rule only stands if we're trying to do it inline, not store it into the variable. 

If we're storing into variable, or pass to another function, user would need to specify explicit `&` / `&mut`, like
```
fun call(r: &mut R) {}
fun main() {
    call(&mut R[@0x1])
}
```

I'm still not sure about the 

```
R[@0x1] = R { value: 1 }
```

It replaces the old R[@0x1] value? Or just reassigns fields?

Reply:
> `R[@0x1]` initially is just a generic "view" on the resource in the global storage. If it's used to read/write it's fields, it becomes either `&` or `&mut` depends on the operation. So one can do:
> 
> ```
> // &R 
> 1 + R[@0x1].value;
> // &mut R 
> R[@0x1].value = 1;
> ```

Yes, according to the current design, if it is used to read/write fields, `&` or `&mut` is automatically inserted, regardless of whether used in an expression or doing something like `let a = R[@0x1].value`.

Reply:
For standalone `R[@0x1]`, it depends on whether it appears on lhs (to update the value) or rhs (to get the value). If we are going to update the value, `R[@0x1] = R {value: 1}` is equivalent to `*(borrow_global_mut<R>(@0x1) = R {value: 1}`.  When appearing on the rhs, `R[@0x1]` is equivalent to `*(borrow_global<R>(@0x1))` to get the actual value.

Reply:
```
(borrow_global_mut<R>(@0x1)) = R {value: 1}
```
I didn't know you can assign resources like that.

Reply:
Still, isn't it too much sugar though, too many implicit rules? I can understand the magic of vector `v[0]` considering that it's restricted only to the items with `copy`, so you can do auto-borrow/auto-deref without issues. 
But here, there's a lot of magic where it's been none before.

Reply:
Not sure whether it is type but you need to use `*(borrow_global_mut<R>(@0x1)) = R {value: 1}` to update the value

Reply:
Thanks for the feedback @mkurnikov! Any comments/suggestions on this?  @wrwg @vineethk